source/common/router/upstream_request.cc

```diff
@@ -538,9 +539,40 @@ void TcpUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
 }
 
 void TcpUpstream::encodeHeaders(const Http::RequestHeaderMap&, bool end_stream) {
-  if (end_stream) {
-    Buffer::OwnedImpl data;
-    upstream_conn_data_->connection().write(data, true);
+  // Headers should only happen once, so use this opportunity to add the proxy
+  // proto header, if configured.
+  ASSERT(upstream_request_->parent().routeEntry()->connectConfig().has_value());
+  Buffer::OwnedImpl data;
+  if (upstream_request_->parent()
+          .routeEntry()
+          ->connectConfig()
+          .value()
+          .has_proxy_protocol_config()) {
+    auto version = upstream_request_->parent()
```

Comment:
I suspect we are going to need this elsewhere so it might be worth it to do a utility that takes the config, the connection, and the buffer to fill, but up to you if you want to do this now or not.